assets/eip-5058/ERC5058.sol

```diff
@@ -0,0 +1,281 @@
+// SPDX-License-Identifier: CC0-1.0
+
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import "./IERC5058.sol";
+
+/**
+ * @dev Implementation ERC721 Lockable Token
+ */
+abstract contract ERC5058 is ERC721, IERC5058 {
+    // Mapping from token ID to unlock time
+    mapping(uint256 => uint256) public lockedTokens;
+
+    // Mapping from token ID to lock approved address
+    mapping(uint256 => address) private _lockApprovals;
+
+    // Mapping from owner to lock operator approvals
+    mapping(address => mapping(address => bool)) private _lockOperatorApprovals;
+
+    /**
+     * @dev See {IERC5058-lockApprove}.
+     */
+    function lockApprove(address to, uint256 tokenId) public virtual override {
+        require(!isLocked(tokenId), "ERC5058: token is locked");
+        address owner = ERC721.ownerOf(tokenId);
+        require(to != owner, "ERC5058: lock approval to current owner");
+
+        require(
+            _msgSender() == owner || isLockApprovedForAll(owner, _msgSender()),
+            "ERC5058: lock approve caller is not owner nor approved for all"
+        );
+
+        _lockApprove(owner, to, tokenId);
+    }
+
+    /**
+     * @dev See {IERC5058-getLockApproved}.
+     */
+    function getLockApproved(uint256 tokenId) public view virtual override returns (address) {
+        require(_exists(tokenId), "ERC5058: lock approved query for nonexistent token");
+
+        return _lockApprovals[tokenId];
+    }
+
+    /**
+     * @dev See {IERC5058-lockerOf}.
+     */
+    function lockerOf(uint256 tokenId) public view virtual override returns (address) {
+        require(_exists(tokenId), "ERC5058: locker query for nonexistent token");
+        require(isLocked(tokenId), "ERC5058: locker query for non-locked token");
+
+        return _lockApprovals[tokenId];
+    }
+
+    /**
+     * @dev See {IERC5058-setLockApprovalForAll}.
+     */
+    function setLockApprovalForAll(address operator, bool approved) public virtual override {
+        _setLockApprovalForAll(_msgSender(), operator, approved);
+    }
+
+    /**
+     * @dev See {IERC5058-isLockApprovedForAll}.
+     */
+    function isLockApprovedForAll(address owner, address operator) public view virtual override returns (bool) {
+        return _lockOperatorApprovals[owner][operator];
+    }
+
+    /**
+     * @dev See {IERC5058-isLocked}.
+     */
+    function isLocked(uint256 tokenId) public view virtual override returns (bool) {
+        return lockedTokens[tokenId] > block.number;
+    }
+
+    /**
+     * @dev See {IERC5058-lockExpiredTime}.
+     */
+    function lockExpiredTime(uint256 tokenId) public view virtual override returns (uint256) {
+        return lockedTokens[tokenId];
+    }
+
+    /**
+     * @dev See {IERC5058-lockFrom}.
+     */
+    function lockFrom(
+        address owner,
```

Comment:
Also, I don't think we need to get owner as a parameter because we can get this info via `ERC721.ownerOf(tokenId)`.